syntax = "proto3";

package abd;

message Tag {
  uint64 counter = 1;   // Lamport-style counter
  string client_id = 2;
}

// Phase 1 of WRITE: client asks for the current tag of a key.
message WriteQueryRequest {
  string key = 1;
}

message WriteQueryReply {
  Tag tag = 1;          // Current tag for this key at this server
}

// Phase 1 of READ: client asks for current (tag, value) of a key.
message ReadQueryRequest {
  string key = 1;
}

message ReadQueryReply {
  Tag tag   = 1;        // Current tag for this key at this server
  bytes value = 2;      // Current value; use string if you prefer
}

// Phase 2 of WRITE and READ: client propagates tagged value.
message WritePropRequest {
  string key = 1;
  Tag tag = 2;
  bytes value = 3;
}

// Generic ACK reply for write propagation.
message Ack {
  bool ok = 1;          // true if server processed the request
  string error = 2;     // optional error message (for logging/debug)
}

message AcquireLockRequest {
  string key = 1;
  string client_id = 2;
}

message AcquireLockReply {
  bool granted = 1;
  string holder = 2;      // optional: tells who holds the lock if not granted
}

message ReleaseLockRequest {
  string key = 1;
  string client_id = 2;
}

message ReleaseLockReply {
  bool ok = 1;
}


// ---------- ABD Service ----------

service ABDService {
  rpc WriteQuery(WriteQueryRequest) returns (WriteQueryReply);

  rpc ReadQuery(ReadQueryRequest) returns (ReadQueryReply);

  rpc WriteProp(WritePropRequest) returns (Ack);

  rpc AcquireLock(AcquireLockRequest) returns (AcquireLockReply);
  rpc ReleaseLock(ReleaseLockRequest) returns (ReleaseLockReply);
}
